### PR TITLE
[move-lang] rewrote alias scoping, improving compiler perf by ~65%

### PR DESCRIPTION
**Pull request by tnowacki**
- Old alias scoping was done with an iter+filter to do set difference
- New alias scoping instead keeps track of unused items at each scope
- For each new scope, the new algorithm fully populates a set of "unused" aliases. Each usage of an alias then removes that from the set, so at the end of the scope, it can instantaneously answer what aliases are unused (as opposed to doing set difference)
- Additionally, clones are reduced when the scope does not introduce anything new. This accounted for about an additional 11% improvement on release builds

## Motivation

- According to flamegraph, expansion was taking 60% to 70% of the total runtime for the compiler
- This motivated the rewrite and led to some nice perf gains!
```
dev build DF+Std (averaged over 10 iterations)
OLD       8.155s
NEW       2.840s
REDUCTION 65.17%

release build DF+Std (averaged over 100 iterations)
OLD       0.849s
NEW       0.271s
REDUCTION 68.08%

nextest -p move-lang
OLD       17.623s
NEW       16.102s
REDUCTION 8.63%

nextest -p move-lang-functional-tests
OLD       511.871s
NEW       185.475s
REDUCTION 63.77%

nextest -p move-prover
OLD       650.349s
NEW       287.221s
REDUCTION 55.84%

x test -p move-lang
OLD       10.047s
NEW       7.335s
REDUCTION 26.99%

x test -p move-lang-functional-tests
OLD       51.223s
NEW       42.450s
REDUCTION 17.13%

x test -p move-prover
OLD       534.380s
NEW       228.259s
REDUCTION 57.29%

```

## Test Plan

- Actually found a bug while doing this too! Updated that test

### CODE REVIEW

File: language/move-lang/src/expansion/aliases.rs [hunk 1]
```diff
@@ -3,64 +3,63 @@
...
-    pub modules: BTreeSet<Name>,
-    pub members: BTreeSet<Name>,
+    pub modules: UniqueSet<Name>,
+    pub members: UniqueSet<Name>,
+}
+
+#[derive(Clone, Debug)]
+pub struct AliasMapBuilder {
+    modules: UniqueMap<Name, (ModuleIdent, /* is_implicit */ bool)>,
+    members: UniqueMap<Name, ((ModuleIdent, Name), /* is_implicit */ bool)>,
```
Comment (modocache) [inline language/move-lang/src/expansion/aliases.rs:21]: I've read a lot of Internet armchair opinions recently on using enums vs. booleans, and while I discount most of the stuff I read on the Internet, I do wonder whether an enum would make more sense here.

Most spots in this file where `true` or `false` is passed into these tuples are also accompanied by a comment, `/* is_implicit */`. An alternative would be an `enum AliasKind { Explicit, Implicit }`, which would allow you to pass in `Implicit` instead. Readers of the code would know what that means, without needing to read a code comment, or add their own if they modify this file.

(And, should an additional alias "kind" be added to the language later, it'd be easier to extend the `enum` than it would be to modify the `bool` use sites.)

File: language/move-lang/src/expansion/aliases.rs [hunk 2]
```diff
@@ -125,78 +123,137 @@ impl AliasMap {
...
+        self.unused.push(AliasSet::new());
+        for tp_name in tparams_iter {
+            self.members.remove(tp_name);
         }
-        self.current_scope = new_scope
+        outer_scope
     }
 
-    pub fn close_scope_and_report_unused(&mut self, inner: Self) -> AliasSet {
+    /// Resets the alias map and gives the set of aliases that were used
```
Comment (modocache) [inline language/move-lang/src/expansion/aliases.rs:237]: Should "used" in the comment here be "unused" instead? I might be misreading but that seems like what the code is doing.

File: language/move-lang/src/expansion/aliases.rs [hunk 3]
```diff
@@ -125,78 +123,137 @@ impl AliasMap {
...
+    pub fn set_to_outer_scope(&mut self, outer_scope: OldAliasMap) -> AliasSet {
+        let outer_scope = match outer_scope.0 {
+            None => return AliasSet::new(),
+            Some(outer) => outer,
+        };
+        let mut inner_scope = std::mem::replace(self, outer_scope);
```
Comment (modocache) [inline language/move-lang/src/expansion/aliases.rs:243]: Wow!! TIL you can do something like this in Rust.

**Review by modocache — APPROVED** [review]
Awesome! I just had one armchair opinion about enums vs. booleans, please take it or leave it 👍

File: language/move-lang/src/expansion/aliases.rs (hunk 3 above)
Comment (tnowacki) [reply language/move-lang/src/expansion/aliases.rs:243]: std::mem is a fun place :)

File: language/move-lang/src/expansion/aliases.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/src/expansion/aliases.rs:21]: I'd definitely agree with you if the flag was public. 
But as it is internal, I don't think the weight of an enum is worth it. The limited usage internally makes this fine for me
```
if !is_implicit { ... }
```
is a lot more ergonomic to me than 
```
if matches(implicitness, AliasKind::Explicit) { ... }
```
And I really hate this because the match isn't explicit... so I would probably end up doing
```
match implicitness {
  AK::Implicit =>  (),

}
```


This would be different if the usages were always 
```
if is_implicit { /* case 1 * } else { /* case 2 */ }
```
since 
```
match implicitness {
  AK::Implicit => /* case 1 */,
  AK::Explicit => /* case 2 */,
}
```
would be much more clear

tldr: it is an internal flag that is only ever used for a single branch so I feel an enum would make things more confusing in many places

Comment (tnowacki) [reply language/move-lang/src/expansion/aliases.rs:21]: So maybe a better summary of I use enums like this:
- Enums are great for flags for public APIs
- Enums are great for flags when you want to do something in each case
- Enums not so great when used for guard-style programming (as I tend to dislike non-exhaustive matches, option is an exception usually since it is always and forever only have two cases)

Comment (modocache) [reply language/move-lang/src/expansion/aliases.rs:21]: Makes sense to me!

**Review by modocache — APPROVED** [review]
